src/mip/cg_strengthening.jl

```diff
@@ -0,0 +1,134 @@
+coefficients and round down negative ones, so this procedure
+will not reduce any coefficient to 0.
+"""
+
+function cg_strengthening!(ps::PresolveData{T}, resolution::Int = 5) where {T}
+    ps.pb0.is_continuous && return nothing
+    for i in 1:ps.pb0.ncon
+        ps.rowflag[i] || continue
+        nzind = ps.pb0.arows[i].nzind
+        nzval = ps.pb0.arows[i].nzval
+
+        if isfinite(ps.urow[i]) && isfinite(ps.lrow[i]) ||
+            any(ps.var_types[[j for j in nzind if ps.colflag[j]]] .== CONTINUOUS)
+            # skip ranged constraints
+            # and any constraint that involves continuous variables
+            continue
+        elseif isfinite(ps.lrow[i])
+            s, newbound = lowerbound_cg_inequality(nzval, nzind, ps.lrow[i],
+                                    ps.colflag, ps.lcol, ps.ucol, resolution)
+            if s != 0
+                ps.lrow[i] = newbound
+                ps.pb0.arows[i].nzval = [aij > 0 ? ceil(s*aij) : floor(s*aij) for aij in nzval]
+            end
+        elseif isfinite(ps.urow[i])
+            s, newbound = lowerbound_cg_inequality(-nzval, nzind, -ps.urow[i],
+                                    ps.colflag, ps.lcol, ps.ucol, resolution)
+            if s != 0
+                ps.urow[i] = -newbound
+                ps.pb0.arows[i].nzval = [aij > 0 ? ceil(s*aij) : floor(s*aij) for aij in nzval]
+            end
+        end
+    end
+    return nothing
+end
+
+function lowerbound_cg_inequality(nzval::Vector{T}, nzind::Vector{Int},
+    b::T, colflag::Vector{Bool},
+    var_lb::Vector{T}, var_ub::Vector{T},
+    resolution::Int) where {T}
+    # Finding the CG inequality for constraints for the form ax ⩾ b,
+    # Return s and new bound, if s = 0, then no CG inequality is found.
+    row_pos_index = Int[]
+    row_neg_index = Int[]
+    col_pos_index = Int[]
+    col_neg_index = Int[]
+    for (k, (j, aij)) in enumerate(zip(nzind, nzval))
+        if aij > 0 && isfinite(var_lb[j])
+            if colflag[j]
+                b -= aij * var_lb[j]
+                push!(row_pos_index, k)
+                push!(col_pos_index, j)
+            end
+        elseif aij < 0 && isfinite(var_ub[j])
+            if colflag[j]
+                b -= aij * var_ub[j]
+                push!(row_neg_index, k)
+                push!(col_neg_index, j)
+            end
+        else
+            # if any positive variable has -inf lower bound
+            # or any negative variable has inf upper bound, then terminate.
+            return 0, b
+        end
+    end
+
+    # building set for heuristicly finding s.
+    a = abs.([aij for (j, aij) in zip(nzind, nzval) if colflag[j]])
+    set = build_heuristic_set(a, resolution)
+
+    # perform search for s and update.
+    s = 0
+    for h in set
+        found_s = false
+        for val in a
+            if val*ceil(b*h) < ceil(h*val)*b
+                found_s = false
+                break
+            elseif val*ceil(b*h) > ceil(h*val)*b
```

Comment:
Logic seems off here: Isn't this loop (as written) equivalent to
```jl
found_s = val * ceil(b * h) != ceil(h * val) * b
```
?

Reply:
I think it not equivalent to what you write. We want to find h such that
val*ceil(b*h) >= ceil(h*val)*b for every val, and at least one of them is strictly greater, i.e,
val*ceil(b*h) > ceil(h*val)*b.

The logic here is that if any val such that val*ceil(b*h) < ceil(h*val)*b, then h is not the value we're looking for, we break the loop right away. If every val satisfies val*ceil(b*h) >= ceil(h*val)*b, then at least one of them must be strictly greater, otherwise h is not what we looking for.

If I set found_s = val * ceil(b * h) != ceil(h * val) * b, then found_s is true everytime val * ceil(b * h) is not equal to ceil(h * val) * b, which is not what we want

Reply:
No you're right, I misread this.